landlab/components/space/ext/calc_qs_v2.pyx

```diff
@@ -0,0 +1,194 @@
+import numpy as np
+cimport numpy as np
+cimport cython
+
+cdef extern from "math.h":
+    double exp(double x) nogil
+    
+from libc.math cimport log,exp
+DTYPE_FLOAT = np.double
+ctypedef np.double_t DTYPE_FLOAT_t
+
+DTYPE_INT = int
+ctypedef np.int_t DTYPE_INT_t
+
+DTYPE_UINT8 = np.uint8
+ctypedef np.uint8_t DTYPE_UINT8_t
+
+# from libc.stdio cimport printf
+
+
+def calculate_qs_in(np.ndarray[DTYPE_INT_t, ndim=1] stack_flip_ud,
+                    np.ndarray[DTYPE_INT_t, ndim=1] flow_receivers,                    
+                    np.ndarray[DTYPE_UINT8_t, ndim=1] node_Status,
+                    np.ndarray[DTYPE_FLOAT_t, ndim=1] cell_area,
+                    np.ndarray[DTYPE_FLOAT_t, ndim=1] q,
+                    np.ndarray[DTYPE_FLOAT_t, ndim=1] qs,
+                    np.ndarray[DTYPE_FLOAT_t, ndim=1] qs_in,
+                    np.ndarray[DTYPE_FLOAT_t, ndim=1] Es,
+                    np.ndarray[DTYPE_FLOAT_t, ndim=1] Er,
+                    np.ndarray[DTYPE_FLOAT_t, ndim=1] Q_to_the_m,
+                    np.ndarray[DTYPE_FLOAT_t, ndim=1] slope,
+                    np.ndarray[DTYPE_FLOAT_t, ndim=1] H,
+                    np.ndarray[DTYPE_FLOAT_t, ndim=1] br,
+                    np.ndarray[DTYPE_FLOAT_t, ndim=1] sed_erosion_term,                    
+                    np.ndarray[DTYPE_FLOAT_t, ndim=1] bed_erosion_term,
+                    DTYPE_FLOAT_t v,
+                    DTYPE_FLOAT_t phi,
+                    DTYPE_FLOAT_t F_f,                    
+                    DTYPE_FLOAT_t K_sed,
+                    DTYPE_FLOAT_t H_star,
+                    DTYPE_FLOAT_t dt,                    
+                    DTYPE_FLOAT_t thickness_lim):
+    
+    """Calculate and qs and qs_in."""
+    # define internal variables
+    cdef unsigned int n_nodes = stack_flip_ud.size
+    cdef unsigned int node_id
+    cdef unsigned int i
+    cdef double H_Before
+    cdef double vol_SSY_riv
+    vol_SSY_riv =0.0
+
+    for i in range(n_nodes):        
+        # choose the node id
+        node_id = stack_flip_ud[i]
+        if node_Status[node_id]==0:
```

Comment:
> ```python
> np.where((node_status == NodeStatus.CORE) & (q > 0.0))
> ```

That is a great suggestion. I slightly adjusted since we need to keep the order of the stack (looping through, to calculate drainage area from highest to lowest cell): 

> ```python
> stack_flip_ud_sel = stack_flip_ud[(node_status[stack_flip_ud] == NodeStatus.CORE) & (self._q[stack_flip_ud] > 0.0)]
> ```